posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+linkTo (Capture rest)    = ??? : linkTo rest
+```
+
+But... what should we put in place of those `???`, if anything?
+
+Well, we definitely want to add _some_ path component, to fill the
+`Capture` slot. However, by definition, a captured path fragment is
+not fixed, it is allowed to vary. In other words, `Capture :> Verb Post`
+matches both `POST /x` and `POST /y`. We cannot just pick one value
+and hope that it is the one the user wanted. We need to take it as an
+argument. But what about `Capture :> Capture :> Verb Post`? We would need
+our `linkTo` function to take 2 arguments for that case. And zero additional
+argument for `Static "hello" :> Verb Post`. This is quite problematic.
+
+Indeed, we would like the type of `linkTo` to be `Endpoint -> Link`,
+`Endpoint -> String -> Link`, `Endpoint -> String -> String -> Link` and
+so on depending on what the `Endpoint` argument is. In other words,
+we want the return type of `linkTo` (when really seen as a function of
+one argument, which it is anyway) to depend on the value of type `Endpoint`
+it gets as input. That is, we want a type that depends on a value, i.e
+dependent types.
+
+The alternative would be:
+
+``` haskell
+-- linkTo assumes that the capture values are given in the same order as we want
+-- them to appear in the url.
+linkTo :: Endpoint -> [String] -> Link
+linkTo (Static str rest) captureValues  = str : linkTo rest captureValues
+-- when there is at least one value left in the list, we use it.
+-- when there isn't... we error out.
+linkTo (Capture rest)    (c : cs)       = c   : linkTo rest cs
+linkTo (Capture rest)    []             = error "linkTo: capture value needed but the list is empty" -- :-(
+linkTo (Verb method)     _              = []
+```
+
+This solution is very unsatisfactory, first and foremost because it is possible
+for it to error out if we don't supply the right number of captures. However,
+it will also silently let us pass too many capture values without telling us
+that some of them are not used. Such a function should be total, we really don't
+want an implementation that can let us down if we're not very careful.
+
+Fortunately, GADTs can help here. We could turn `Endpoint` into
+a GADT that tracks captures and then use some type-level computations
+to get the type of the link-making function from our list of captures, as well
+as define the link making function through typeclass instances that would
+go through the captures and add an argument for each of them. Request bodies,
+query parameters, headers? We could probably track them too, in a similar way.
+Or we could unify it all by basically building up and tracking actual servant
+API types through a `GADT` version of Endpoint's type argument, and do some of
+what servant does at the type-level, with everything else at the value-level.
+
+However, all those approaches have a big problem. Once you've made a decision,
+it is set in stone, in a way. To be more precise, with all these approaches,
```

Comment:
> To be more precise...

Very long sentence,  hard to grok, but it is important.

Reply:
Splitting it up in 2 sentences.